polar-core/src/resource_block.rs

```diff
@@ -280,21 +282,61 @@ impl ResourceBlocks {
         resource: Term,
         declarations: Declarations,
         shorthand_rules: Vec<ShorthandRule>,
-    ) {
-        self.declarations.insert(resource.clone(), declarations);
-        self.shorthand_rules
-            .insert(resource.clone(), shorthand_rules);
+    ) -> Vec<ValidationError> {
+        let mut errors = vec![];
+
+        // Merge existing declarations if we are reopening a resource block, otherwise add new
+        if let Some(existing) = self.declarations.get_mut(&resource) {
+            for (key, new) in &declarations {
+                let existing_declaration = existing.get(key);
+                // allow
+                // 1. overwriting declarations of the **same** type
+                // 2. inserting declarations where none previously existed
+                //
+                // deny overwriting of declarations with mismatched types
+                match (new, existing_declaration) {
+                    // new or same-type declarations are accepted
+                    (Declaration::Relation(_), Some(Declaration::Relation(_)))
+                    | (Declaration::Relation(_), None)
+                    | (Declaration::Role, Some(Declaration::Role))
+                    | (Declaration::Role, None)
+                    | (Declaration::Permission, Some(Declaration::Permission))
+                    | (Declaration::Permission, None) => {
+                        existing.insert(key.clone(), new.clone());
+                    }
+                    // disallow all other combinations (mismatched declaration types)
+                    _ => {
+                        errors.push(ValidationError::DuplicateResourceBlockDeclaration {
+                            resource: resource.clone(),
+                            declaration: key.clone(),
+                            existing: existing_declaration.unwrap().clone(),
+                            new: new.clone(),
+                        });
+                    }
+                }
+            }
+        } else {
+            // or insert a new set of declarations for the resource type
+            self.declarations.insert(resource.clone(), declarations);
+        }
+
+        // Merge existing shorthand rules if we are reopening a resource block, otherwise add new
+        if let Some(existing) = self.shorthand_rules.get_mut(&resource) {
+            existing.extend(shorthand_rules.into_iter());
+        } else {
+            self.shorthand_rules
+                .insert(resource.clone(), shorthand_rules);
+        }
```

Comment:
This is a fine way to write this (and avoids an unnecessary `.clone()` in the "merge" case), but just FYI in case you haven't encountered [the Entry API](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html) yet:

```rust
        self.shorthand_rules
            .entry(resource.clone())
            .and_modify(|existing| existing.extend(shorthand_rules.clone()))
            .or_insert(shorthand_rules);
```

Reply:
TIL about the Entry API 🎉 updated to use the chained form